crates/shared/src/price_estimation/buffered.rs

```diff
@@ -0,0 +1,459 @@
+//! A buffered implementation that automatically groups native prices API
+//! requests into batches.
+
+use {
+    crate::price_estimation::{native::NativePriceEstimateResult, PriceEstimationError},
+    anyhow::anyhow,
+    async_trait::async_trait,
+    futures::{
+        channel::mpsc,
+        future::FutureExt as _,
+        stream::{self, FusedStream, Stream, StreamExt as _},
+    },
+    primitive_types::H160,
+    std::{
+        collections::{HashMap, HashSet},
+        future::Future,
+        num::NonZeroUsize,
+        sync::Arc,
+        time::Duration,
+    },
+    tokio::{
+        sync::broadcast,
+        task::JoinHandle,
+        time::{error::Elapsed, sleep},
+    },
+};
+
+/// Buffered configuration.
+#[derive(Clone)]
+pub struct Configuration {
+    /// The maximum amount of concurrent batches to request.
+    ///
+    /// Specifying `None` means no limit on concurrency.
+    pub max_concurrent_requests: Option<NonZeroUsize>,
+    /// The maximum batch size.
+    pub max_batch_len: usize,
+    /// An additional minimum delay to wait for collecting requests.
+    ///
+    /// The delay starts counting after receiving the first request.
+    pub batch_delay: Duration,
+}
+
+/// Trait for fetching a batch of native price estimates.
+#[mockall::automock]
+#[async_trait]
+pub trait NativePriceBatchFetcher: Sync + Send {
+    /// Fetches a batch of native price estimates.
+    ///
+    /// It returns a HashMap which maps the token with its price
+    async fn fetch_native_prices(
+        &self,
+        tokens: &HashSet<H160>,
+    ) -> Result<HashMap<H160, f64>, PriceEstimationError>;
+}
+
+/// Buffered implementation that implements automatic batching of
+/// native prices requests.
+#[derive(Clone)]
+pub struct BufferedTransport<Inner> {
+    config: Configuration,
+    #[allow(dead_code)]
+    inner: Arc<Inner>,
+    calls: mpsc::UnboundedSender<H160>,
+    broadcast_sender: broadcast::Sender<NativePriceResult>,
+}
+
+type NativePriceResult = (H160, Result<f64, PriceEstimationError>);
+
+impl<Inner> BufferedTransport<Inner>
+where
+    Inner: NativePriceBatchFetcher + Send + Sync + 'static,
+{
+    /// Maximum capacity of the broadcast channel, the messages are discarded as
+    /// soon as they are sent, so this limit should be enough to hold the
+    /// flow
+    const BROADCAST_CHANNEL_CAPACITY: usize = 50;
+
+    /// Creates a new buffered transport with the specified configuration.
+    pub fn with_config(inner: Inner, config: Configuration) -> Self {
+        let inner = Arc::new(inner);
+        let (calls, receiver) = mpsc::unbounded();
+
+        let (broadcast_sender, _) = broadcast::channel(Self::BROADCAST_CHANNEL_CAPACITY);
```

Comment:
We do not need the broadcast receiver here as it doesn't implement `Clone` and we can get many receivers by just using `tx.subscribe()`.

Reply:
@MartinquaXD Well, the receiver depends on the perspective. So, we have:

- `BufferedTransport` which receives token requests (via `mpsc` channel)
- let's call it `client`, the entity/piece of code requesting the native price

So, there is only one `BufferedTransport` receiving token price requests, this entity batches the petitions the same way we do in `ethrpc` (either by time limit or by size limit, depending which it reached sooner).

This communication is done with the `mpsc` channel.

But, the `client` still needs to get the price back. The most efficient way (memory and performance) is to do a broadcast channel. If you are the client who just requested a price, then you subscribe to the broadcast waiting for your price to come (so, the `client` is the receiver here).

I don't know if I explained myself properly.